spacy/tests/test_visualization.py

```diff
@@ -0,0 +1,1080 @@
+        "<╣ ║",
+        "═╝<╣",
+        "<══╣",
+        "<══╝",
+    ]
+    dep_tree = Visualizer.render_dependency_tree(doc[3 : len(doc)], False)
+    assert dep_tree == [
+        "  ╔>",
+        "  ╠>",
+        "╔>╚═",
+        "╠═══",
+        "║ ╔>",
+        "║ ╠>",
+        "╠>╚═",
+        "╠══>",
+        "╚══>",
+    ]
+
+
+def test_visualization_dependency_tree_non_projective(en_vocab):
+    """Test dependency tree display with a non-projective dependency."""
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+    dep_tree = Visualizer.render_dependency_tree(doc[0 : len(doc)], True)
+    assert dep_tree == [
+        "<╗    ",
+        "═╩═══╗",
+        "<╗   ║",
+        "═╩═╗<╣",
+        "<══║═╣",
+        "<╗ ║ ║",
+        "<╣ ║ ║",
+        "═╝<╝ ║",
+        "<════╝",
+    ]
+    dep_tree = Visualizer.render_dependency_tree(doc[0 : len(doc)], False)
+    assert dep_tree == [
+        "    ╔>",
+        "╔═══╩═",
+        "║   ╔>",
+        "╠>╔═╩═",
+        "╠═║══>",
+        "║ ║ ╔>",
+        "║ ║ ╠>",
+        "║ ╚>╚═",
+        "╚════>",
+    ]
+
+
+def test_visualization_dependency_tree_input_not_span(en_vocab):
+    """Test dependency tree display behaviour when the input is not a Span."""
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+
+    with pytest.raises(AssertionError):
+        Visualizer.render_dependency_tree(doc[1:3], True)
+
+
+def test_visualization_dependency_tree_highly_nonprojective(en_vocab):
+    """Test a highly non-projective tree (colloquial Polish)."""
+    doc = Doc(
+        en_vocab,
+        words=[
+            "Owczarki",
+            "przecież",
+            "niemieckie",
+            "zawsze",
+            "wierne",
+            "są",
+            "bardzo",
+            ".",
+        ],
+        heads=[5, 5, 0, 5, 5, None, 4, 5],
+        deps=["dep"] * 8,
+    )
+    dep_tree = Visualizer.render_dependency_tree(doc[0 : len(doc)], True)
+    assert dep_tree == [
+        "═╗<╗",
+        " ║<╣",
+        "<╝ ║",
+        "<══╣",
+        "═╗<╣",
+        "═══╣",
+        "<╝ ║",
+        "<══╝",
+    ]
+    dep_tree = Visualizer.render_dependency_tree(doc[0 : len(doc)], False)
+    assert dep_tree == [
+        "╔>╔═",
+        "╠>║ ",
+        "║ ╚>",
+        "╠══>",
+        "╠>╔═",
+        "╠═══",
+        "║ ╚>",
+        "╚══>",
+    ]
+
+
+def test_visualization_render_native_attribute_int(en_vocab):
+    doc = Doc(
```

Comment:
If this exact same `doc` is being used across tests, you could define it as a fixture on top.

Reply:
Also, maybe some tests can be merged? It's fine to test a few related things in one bigger test.